src/Microsoft.Tye.Core/ApplicationFactory.cs

```diff
@@ -74,6 +77,87 @@ public static async Task<ApplicationBuilder> CreateAsync(OutputContext output, F
+                // This batching will mitigate the performance cost of running MSBuild out of process.
+                var projectServices = services.Where(s => !string.IsNullOrEmpty(s.Project));
+                var projectMetadata = new Dictionary<string, string>();
+
+                using (var directory = TempDirectory.Create())
+                {
+                    var projectPath = Path.Combine(directory.DirectoryPath, Path.GetRandomFileName() + ".proj");
+
+                    var sb = new StringBuilder();
+                    sb.AppendLine("<Project>");
+                    sb.AppendLine("    <ItemGroup>");
+
+                    foreach (var project in projectServices)
+                    {
+                        var expandedProject = Environment.ExpandEnvironmentVariables(project.Project!);
+                        project.ProjectFullPath = Path.Combine(config.Source.DirectoryName!, expandedProject);
+
+                        if (!File.Exists(project.ProjectFullPath))
+                        {
+                            throw new CommandException($"Failed to locate project: '{project.ProjectFullPath}'.");
+                        }
+
+                        sb.AppendLine($"        <MicrosoftTye_ProjectServices " +
+                            $"Include=\"{project.ProjectFullPath}\" " +
+                            $"Name=\"{project.Name}\" " +
+                            $"BuildProperties=\"{(project.BuildProperties.Any() ? project.BuildProperties.Select(kvp => $"{kvp.Name}={kvp.Value}").Aggregate((a, b) => a + ";" + b) : string.Empty)}\" />");
+                    }
+                    sb.AppendLine(@"    </ItemGroup>");
+
+                    sb.AppendLine($@"    <Target Name=""MicrosoftTye_EvaluateProjects"">");
+                    sb.AppendLine($@"        <MsBuild Projects=""@(MicrosoftTye_ProjectServices)"" "
+                        + $@"Properties=""%(BuildProperties);"
+                        + $@"MicrosoftTye_ProjectName=%(Name)"" "
+                        + $@"Targets=""MicrosoftTye_GetProjectMetadata"" BuildInParallel=""true"" />");
+
+                    sb.AppendLine("    </Target>");
+                    sb.AppendLine("</Project>");
+                    File.WriteAllText(projectPath, sb.ToString());
+
+                    output.WriteDebugLine("Restoring and evaluating projects");
+
+                    var msbuildEvaluationResult = await ProcessUtil.RunAsync(
+                        "dotnet",
+                        $"build " +
+                            $"\"{projectPath}\" " +
+                            $"/p:CustomAfterMicrosoftCommonTargets={Path.Combine(Path.GetDirectoryName(Assembly.GetExecutingAssembly().Location)!, "ProjectEvaluation.targets")} " +
+                            $"/nologo",
+                        throwOnError: false,
+                        workingDirectory: directory.DirectoryPath);
+
+                    // If the build fails, we're not really blocked from doing our work.
```

Comment:
What can cause the build to fail here? Would the failure to evaluate a project be shown when we throw a CommandException?

Reply:
I didn't run into any cases where this fails but to be defensive, I've added the details of the output to the debug logs.